### PR TITLE
Php 7.2 / Phpunit 9 / PSR-4 / Refacto tests

### PR DESCRIPTION
**Pull request by VincentLanglet**
This PR:
- Use PSR4 for the project
- Drop Support < 7.2 since it's not maintained anymore by php
- Bump Phpunit to 8 | 9.
- Clean tests structure.

This is BC. 
- src classes keep the same Namespace
- version is resolved by composer

### CODE REVIEW

**Comment by monofone** [comment]
Just WOW, great you took the time to make some long needed cleanups. I would like to test if this just works in an application of mine. But I must admit thnat this will take a cpuple of days. So please be patient I will take care abut.

I'm uncertain about the strict requirement of the php version in the composer json, I don't get statistics from packagist to see if someone still uses lower php versions, on the other side they can  use an older version.

Yeah ok, I take a look, i hope I get it done to the end of the week.

Regards

**Comment by VincentLanglet** [comment]
> Just WOW, great you took the time to make some long needed cleanups.

I'm happy you take it like this. :)

> I would like to test if this just works in an application of mine. But I must admit thnat this will take a couple of days. So please be patient I will take care abut.

Sure, no problem.

> I'm uncertain about the strict requirement of the php version in the composer json, I don't get statistics from packagist to see if someone still uses lower php versions, on the other side they can use an older version.

php 7.2 is a requirement for phpunit 8.
Keeping the support for old version was a blocking point.

They can still use older version, so I think this should be fine.


> Yeah ok, I take a look, i hope I get it done to the end of the week.

👌

**Comment by VincentLanglet** [comment]
@monofone Hi, did you have time to test it ?

**Comment by monofone** [comment]
Hey,

I'm really sorry, I hope to get it done the next few days, at the moment there is too much on my list.

June 15, 2020 12:35 PM, "Vincent Langlet" <notifications@github.com (mailto:notifications@github.com?to=%22Vincent%20Langlet%22%20<notifications@github.com>)> wrote:
	@monofone (https://github.com/monofone) Hi, did you have time to test it ?

File: tests/Unit/TransferInformation/CustomerCreditTransferInformationTest.php [hunk 1]
```diff
@@ -1,17 +1,18 @@
...
-class CustomerCreditTransferInformationTest extends \PHPUnit_Framework_TestCase
+class CustomerCreditTransferInformationTest extends TestCase
 {
     /**
      * Tests whether the EndToEndId equals the name if no other identifier was supplied
      */
     public function testEndToEndIndentifierEqualsName()
     {
-        $information = new CustomerCreditTransferInformation(100, 'DE12500105170648489890', 'Their Corp');
+        $information = new CustomerCreditTransferInformation('100', 'DE12500105170648489890', 'Their Corp');
```
Comment (monofone) [inline tests/Unit/TransferInformation/CustomerCreditTransferInformationTest.php:15]: the change of the amount to string is not ok as this is not the interface used up to now, the test should also work woth an int or a float in this place

File: tests/Unit/TransferInformation/CustomerCreditTransferInformationTest.php [hunk 2]
```diff
@@ -20,7 +21,7 @@ public function testEndToEndIndentifierEqualsName()
      */
     public function testOptionalEndToEndIdentifier()
     {
-        $information = new CustomerCreditTransferInformation(100, 'DE12500105170648489890', 'Their Corp', 'MyEndToEndId');
+        $information = new CustomerCreditTransferInformation('100', 'DE12500105170648489890', 'Their Corp', 'MyEndToEndId');
```
Comment (monofone) [inline tests/Unit/TransferInformation/CustomerCreditTransferInformationTest.php:24]: see before

File: src/TransferInformation/BaseTransferInformation.php [hunk 3]
```diff
@@ -104,19 +104,14 @@ class BaseTransferInformation implements TransferInformationInterface
...
     public function __construct($amount, $iban, $name)
     {
-        $amount += 0;
-        if (is_float($amount)) {
-            if (!function_exists('bcscale')) {
-                throw new InvalidArgumentException('Using floats for amount is only possible with bcmath enabled');
-            }
-            bcscale(2);
-            $amount = (integer)bcmul(sprintf('%01.4F', $amount), '100');
+        if(preg_match("#^([0-9]+)\.([0-9]+)$#", $amount, $matches)) {
```
Comment (monofone) [inline src/TransferInformation/BaseTransferInformation.php:110]: We need to implement some test for this. This is a huge change in my opinion and it should be safeguarded by some tests

File: tests/Unit/TransferInformation/CustomerCreditTransferInformationTest.php (hunk 1 above)
Comment (VincentLanglet) [reply tests/Unit/TransferInformation/CustomerCreditTransferInformationTest.php:15]: According to the phpDoc, the amount should be a string.
https://github.com/php-sepa-xml/php-sepa-xml/blob/8f9814bc7c97eda669e93848df3dcdb28ca7debb/lib/Digitick/Sepa/TransferInformation/CustomerCreditTransferInformation.php#L28

When you use static analysis (like I do on my project) it require to pass a string. So this should be the tested value.

I can update the phpDoc to `string|float|int` and test the three value if you want. I think it's the best to do.

Comment (monofone) [reply tests/Unit/TransferInformation/CustomerCreditTransferInformationTest.php:15]: Hehe, good catch.

I think we should update the docs. I'd like to use typehints but this won't work here.

Can you tell more about you static analysis I would be interested in what tool you are running.

Regards

Comment (VincentLanglet) [reply tests/Unit/TransferInformation/CustomerCreditTransferInformationTest.php:15]: I updated the doc and added tests then.

I'm using https://github.com/phpstan/phpstan and I love it.
I heard about psalm too but never tried.

**Comment by VincentLanglet** [comment]
@monofone When I wrote the tests, I find kinda confusing that
- `5` means 5 cents.
- `5.0` means 500 cents.

I start to think it's not a good behavior that the two following lines give different results.
```
new BaseTransferInformation((int) $amount, $iban, $name)
new BaseTransferInformation((float) $amount, $iban, $name)
```

Maybe it's better to drop the float support (and release a new major version of course).
We can still provide a service in another class to convert a float in a cents amount.

**Comment by monofone** [comment]
That's exactly what I had as goal for future development. I didn't touched this because of this strange behaviour and the need to create a new major version ;-)

I'm with you lets create a new major.

**Comment by monofone** [comment]
I was at the point that I think it's the responsibility of the application code to provide the value in cents. This is a library to print xml and not to calculate values, specially floats.

**Comment by VincentLanglet** [comment]
Should I drop the float support in this PR ?
Or you prefer to test it and merge it first ; and then I create a new PR with the drop of the float support ?

**Comment by monofone** [comment]
No let's drop it it gets together with my last comment, that the application code should take care

**Comment by VincentLanglet** [comment]
Since we're doing a new major, do you want that I add typehint then ?
Should I restrict the value to `int` or do you want to still allow a `string` ?

**Comment by monofone** [comment]
good, question I think int is totally ok

**Comment by VincentLanglet** [comment]
And do you want that I add typehint for every methods ?

**Comment by monofone** [comment]
you just created a smile on my face, yeah if you will invest this efford I will appreciate it a lot.

**Comment by monofone** [comment]
I'll try to update the docs in the meantime and create a changelog

**Comment by VincentLanglet** [comment]
@monofone Done ! :)

**Comment by VincentLanglet** [comment]
@monofone I saw you made a pre-release.
I you find something wrong, don't hesitate to ping if I can help :)

**Comment by VincentLanglet** [comment]
@monofone Does the new major works for you ? When do you plan to release it ? :)

**Comment by monofone** [comment]
Hi @VincentLanglet ,

holiday season ... I have no plans about releasing it. I would like to take a look into the codeclimate issues and adapt the rules or resolve some of the issues.
And I have to improve my app using this lib to have a better understanding of some issues arrised. (mostly about the float stuff. But this is in my app and unrelated to the lib)

**Comment by VincentLanglet** [comment]
Hi @monofone Any news about the major release ? Do you need help ? :)

**Pull request merged**